node/core/provisioner/src/lib.rs

```diff
@@ -0,0 +1,222 @@
+	const STOP: Self = Self::Stop;
+
+	fn relay_parent(&self) -> Option<Hash> {
+		match self {
+			Self::Provisioner(pm) => pm.relay_parent(),
+			Self::Stop => None,
+		}
+	}
+}
+
+impl TryFrom<AllMessages> for ToJob {
+	type Error = ();
+
+	fn try_from(msg: AllMessages) -> Result<Self, Self::Error> {
+		match msg {
+			AllMessages::Provisioner(pm) => Ok(Self::Provisioner(pm)),
+			_ => Err(()),
+		}
+	}
+}
+
+impl From<ProvisionerMessage> for ToJob {
+	fn from(pm: ProvisionerMessage) -> Self {
+		Self::Provisioner(pm)
+	}
+}
+
+pub enum FromJob {
+	Runtime(RuntimeApiMessage),
+}
+
+impl From<FromJob> for AllMessages {
+	fn from(from_job: FromJob) -> AllMessages {
+		match from_job {
+			FromJob::Runtime(ram) => AllMessages::RuntimeApi(ram),
+		}
+	}
+}
+
+#[derive(Debug, derive_more::From)]
+pub enum Error {
+	#[from]
+	Sending(mpsc::SendError),
+	OneshotReceiverDropped,
+}
+
+impl JobTrait for ProvisioningJob {
+	type ToJob = ToJob;
+	type FromJob = FromJob;
+	type Error = Error;
+	type RunArgs = ();
+
+	const NAME: &'static str = "ProvisioningJob";
+
+	/// Run a job for the parent block indicated
+	//
+	// this function is in charge of creating and executing the job's main loop
+	fn run(
+		relay_parent: Hash,
+		_run_args: Self::RunArgs,
+		receiver: mpsc::Receiver<ToJob>,
+		sender: mpsc::Sender<FromJob>,
+	) -> Pin<Box<dyn Future<Output = Result<(), Self::Error>> + Send>> {
+		async move {
+			let job = ProvisioningJob::new(relay_parent, sender, receiver);
+
+			// it isn't necessary to break run_loop into its own function,
+			// but it's convenient to separate the concerns in this way
+			job.run_loop().await
+		}
+		.boxed()
+	}
+}
+
+impl ProvisioningJob {
+	pub fn new(relay_parent: Hash, sender: mpsc::Sender<FromJob>, receiver: mpsc::Receiver<ToJob>) -> Self {
+		Self {
+			relay_parent,
+			sender,
+			receiver,
+			provisionable_data_channels: Vec::new(),
+			backed_candidates: Vec::new(),
+			signed_bitfields: Vec::new(),
+		}
+	}
+
+	async fn run_loop(mut self) -> Result<(), Error> {
+		while let Some(msg) = self.receiver.next().await {
+			use ProvisionerMessage::{RequestBlockAuthorshipData, RequestInherentData, ProvisionableData};
+
+			match msg {
+				ToJob::Provisioner(RequestInherentData(_, sender)) => self.send_inherent_data(sender).await?,
+				ToJob::Provisioner(RequestBlockAuthorshipData(_, sender)) => self.provisionable_data_channels.push(sender),
+				ToJob::Provisioner(ProvisionableData(data)) => {
+					for channel in self.provisionable_data_channels.iter_mut() {
+						// REVIEW: the try operator here breaks the run loop if any receiver ever unexpectedly
+						// closes their channel. Is that desired?
+						channel.send(data.clone()).await?;
+					}
+					self.note_provisionable_data(data);
+				}
+				ToJob::Stop => break,
+			}
+		}
+
+		Ok(())
+	}
+
+	// REVIEW: not async, but small; hopefully permissable
+	fn note_provisionable_data(&mut self, provisionable_data: ProvisionableData) {
```

Comment:
is "not async" supposed to be a bad thing?

I actually would prefer to always limit the amount of async code to the smallest possible portion, and everything else should be done by a synchronous state machine.

Reply:
When it's called from within async code... yeah, kind of? 

As I understand it, part of the reason that async is complicated to implement is that it inserts suspension points throughout. This makes single-threaded concurrency possible, but it also means that there's no kind of os-style interrupt suspension going on in normal functions; this concurrency is all (in theory) cooperative. This in turn means that if you call a heavy non-async function from within an async function, you either want to use some kind of `spawn_blocking` api, or just deal with the fact that the non-async function will take a while and may starve other async functions which otherwise would be interrupting in the meantime.

As I said in the comment, this function is small and should be pretty lightweight, so the actual impact on the code should be minimal. Still, I did think it was worth specifically mentioning.

Reply:
Well, even when you write async code the work has to happen somewhere. I think the "blocking" work is not intended for stuff like updating hash-maps in memory but more for reading from disk and things like that.